cluster/shard_mapper.go

```diff
@@ -27,56 +26,81 @@ type ShardMapper struct {
 	TSDBStore interface {
 		CreateMapper(shardID uint64, query string, chunkSize int) (tsdb.Mapper, error)
 	}
+
+	timeout time.Duration
+	pool    *clientPool
 }
 
 // NewShardMapper returns a mapper of local and remote shards.
-func NewShardMapper() *ShardMapper {
-	return &ShardMapper{}
+func NewShardMapper(timeout time.Duration) *ShardMapper {
+	return &ShardMapper{
+		pool:    newClientPool(),
+		timeout: timeout,
+	}
 }
 
 // CreateMapper returns a Mapper for the given shard ID.
-func (r *ShardMapper) CreateMapper(sh meta.ShardInfo, stmt string, chunkSize int) (tsdb.Mapper, error) {
+func (s *ShardMapper) CreateMapper(sh meta.ShardInfo, stmt string, chunkSize int) (tsdb.Mapper, error) {
 	var err error
 	var m tsdb.Mapper
-	if sh.OwnedBy(r.MetaStore.NodeID()) {
-		m, err = r.TSDBStore.CreateMapper(sh.ID, stmt, chunkSize)
+	if sh.OwnedBy(s.MetaStore.NodeID()) && !s.ForceRemoteMapping {
+		m, err = s.TSDBStore.CreateMapper(sh.ID, stmt, chunkSize)
 		if err != nil {
 			return nil, err
 		}
 	} else {
-		rm := NewRemoteMaper(sh.OwnerIDs[0], sh.ID, stmt, chunkSize)
-		rm.MetaStore = r.MetaStore
+		conn, err := s.dial(sh.OwnerIDs[0])
```

Comment:
I'd randomize this instead if not using the balancer yet.  Hard-coding to `0` could create a hot spot on one node.